features/exploredash/src/main/java/org/dash/wallet/features/exploredash/AppExploreDatabase.java

```diff
@@ -0,0 +1,175 @@
+import org.dash.wallet.features.exploredash.data.model.Merchant;
+import org.dash.wallet.features.exploredash.data.model.MerchantFTS;
+import org.dash.wallet.features.exploredash.repository.ExploreRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+
+@Database(entities = {
+        Merchant.class,
+        MerchantFTS.class,
+        Atm.class,
+        AtmFTS.class
+}, version = 1)
+@TypeConverters({RoomConverters.class})
+public abstract class AppExploreDatabase extends RoomDatabase {
+
+    private static final Logger log = LoggerFactory.getLogger(AppExploreDatabase.class);
+
+    private static AppExploreDatabase instance;
+
+    public abstract MerchantDao merchantDao();
+
+    public abstract AtmDao atmDao();
+
+    public static AppExploreDatabase getAppDatabase(
+        Context context,
+        Configuration config,
+        ExploreRepository repository
+    ) {
+        if (instance == null) {
+            instance = create(context, config, repository);
+        }
+        return instance;
+    }
+
+    private static AppExploreDatabase create(
+        Context context,
+        Configuration config,
+        ExploreRepository repository
+    ) {
+        String exploreDatabaseName = config.getExploreDatabaseName();
+
+        File dbFile = context.getDatabasePath(exploreDatabaseName);
+        File dbUpdateFile = repository.getUpdateFile();
+        if (!dbFile.exists() && !dbUpdateFile.exists()) {
+            repository.preloadFromAssetsInto(dbUpdateFile);
+        }
+
+        if (dbUpdateFile.exists()) {
+            log.info("found explore db update package {}", dbUpdateFile.getAbsolutePath());
+
+            repository.deleteOldDB(dbFile);
+
+            long dbTimestamp = repository.getTimestamp(dbUpdateFile);
+            exploreDatabaseName = config.setExploreDatabaseName(dbTimestamp);
+        }
+
+        Builder<AppExploreDatabase> dbBuilder = Room.databaseBuilder(
+                context,
+                AppExploreDatabase.class,
+                exploreDatabaseName
+        );
+
+        log.info("Build database {}", exploreDatabaseName);
+        AppExploreDatabase database = buildDatabase(dbBuilder, repository, dbUpdateFile);
+
+        if (!dbFile.exists()) {
+            // execute simple query to trigger database opening (onOpenPrepackagedDatabase)
+            database.query("SELECT * FROM sqlite_master", null);
+        }
+
+        return database;
+    }
+
+    @VisibleForTesting
+    public static AppExploreDatabase buildDatabase(
+        Builder<AppExploreDatabase> dbBuilder,
+        ExploreRepository repository,
+        File dbUpdateFile
+    ) {
+        if (dbUpdateFile.exists()) {
+            log.info("create explore db from InputStream");
+            dbBuilder.createFromInputStream(
+                    () -> repository.getDatabaseInputStream(dbUpdateFile),
+                    new PrepackagedDatabaseCallback() { });
```

Comment:
The update is finalized in the `onOpen` callback instead of here because here the database still appears empty even if it was prepopulated.